ScopedStorage/app/src/main/java/com/samples/storage/saf/SafFragmentViewModel.kt

```diff
@@ -0,0 +1,66 @@
+package com.samples.storage.saf
+
+import android.Manifest.permission.MANAGE_EXTERNAL_STORAGE
+import android.content.ContentResolver
+import android.content.Intent
+import android.net.Uri
+import android.provider.DocumentsContract
+import androidx.documentfile.provider.DocumentFile
+import androidx.lifecycle.ViewModel
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.io.InputStream
+import java.io.OutputStream
+import java.security.MessageDigest
+
+/**
+ * ViewModel contains various examples for how to work with the contents of documents
+ * opened with the Storage Access Framework.
+ */
+class SafFragmentViewModel : ViewModel() {
+    /**
+     * Simple example of using [DocumentFile] to get all the documents in a folder (by using
+     * [Intent.ACTION_OPEN_DOCUMENT_TREE]).
+     * It's possible to use [DocumentsContract] and [ContentResolver] directly, but using
+     * [DocumentFile] allows us to access an easier to use API.
+     *
+     * While it's _possible_ to search across multiple directories and recursively work with files
+     * via SAF, there can be significant performance penalties to this type of usage. If your
+     * use case requires this, consider looking into the permission [MANAGE_EXTERNAL_STORAGE].
+     *
+     * Accessing any field in the [DocumentFile] object, aside from [DocumentFile.getUri],
+     * ultimately performs a lookup with the system's [ContentResolver], and should thus be
+     * performed off the main thread, which is why we're doing this transformation from
+     * [DocumentFile] to file name and [Uri] in a coroutine.
+     */
+    suspend fun listFiles(folder: DocumentFile): List<Pair<String, Uri>> {
+        return if (folder.isDirectory) {
```

Comment:
`DocumentFile.isDirectory()` involves IPC and should be called from a background thread.